src/tools/schema2base.ts

```diff
@@ -31,6 +31,12 @@ export interface ClassGenerator {
   generate(schemaHash: string, fieldCount: number): string;
 }
 
+export class NodeAndGenerator {
+  generator: ClassGenerator;
+  hash: string;
+  fieldLength: number;
+}
```

Comment:
Not sure this is the best approach. My initial thought for this was to pass in a language-specific callback that creates the names and aliases for that langauge.

So you wouldn't need `kotlinName` and `kotlinAliases`, because the standard name and aliases fields would be populated using the Schema2Kotlin-provided logic. Does that make sense?

Reply:
Yes, that makes sense and I'm happy to do so, but I think it should be in a separate PR. The main purpose of this PR is to move the entity classes within the abstract particle classes, and that is the purpose of NodeAndGenerator.

schema2base creates the entity classes and then the particle classes. But we need the entity classes to go inside the particle classes. In other words, for Kotlin we now want the order to be GenerateTypeAliases -> GenerateAbstractParticle (including entity classes) but C++ should still be GenerateEntityClasses (including TypeAliases) -> GenerateAbstractParticle.

NodeAndGenerator allows us to provide the information required for the particle generator to create the entity classes while still having schema2base call the functions in the same order for C++ and Kotlin.

I think this is a bit hack-y, but I worked with it for quite awhile and it seemed to be a minimally invasive method to the challenge.  I'm open to other suggestions. Does that all make sense?

Reply:
I'm ok with you fixing this in a follow up, so that you can land this before everything gets refactored again ;-)

Reply:
Agreed, happy to go with this and revisit later.